### PR TITLE
Yogi

### LINKED ISSUE #9: Implement support for Yogi optimiser
See paper:

https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization.pdf

### PR DESCRIPTION
**Pull request by joaogui1**
Closes #9

### CODE REVIEW

File: optax/_src/transform.py [hunk 1]
```diff
@@ -408,6 +408,47 @@ def update_fn(updates, state, params=None):
...
+
+  References:
+    [Zaheer et al, 2018](https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization)
+
+  Args:
+    b1: decay rate for the exponentially weighted average of grads.
+    b2: decay rate for the exponentially weighted average of variance of grads.
+    eps: term added to the denominator to improve numerical stability.
+    eps_root: term added to the denominator inside the square-root to improve
+      numerical stability when backpropagating gradients through the rescaling.
```
Comment (mtthss) [inline optax/_src/transform.py:425]: Missing `initial_accumulator_value` in `Args`

File: optax/_src/transform.py [hunk 2]
```diff
@@ -408,6 +408,47 @@ def update_fn(updates, state, params=None):
...
+  def init_fn(params):
+    value_like = lambda p: jnp.full_like(p, initial_accumulator_value)
+    mu = jax.tree_map(value_like, params)  # First moment
+    nu = jax.tree_map(value_like, params)  # Second Central moment
+    return ScaleByAdamState(count=jnp.zeros([], jnp.int32), mu=mu, nu=nu)
+
+  def update_fn(updates, state, params=None):
+    del params
+    mu = _update_moment(updates, state.mu, b1, 1)
+    signed_sq = jax.tree_multimap(lambda g, v: jnp.sign(v - g**2)*g**2, updates, state.nu)
```
Comment (mtthss) [inline optax/_src/transform.py:440]: Please break line 440 after `tree_multimap(`

File: optax/_src/transform.py [hunk 3]
```diff
@@ -408,6 +408,47 @@ def update_fn(updates, state, params=None):
...
 
 
+def scale_by_yogi(
+    b1: float = 0.9, b2: float = 0.999,
+    eps: float = 1e-3, eps_root: float = 0.0, 
+    initial_accumulator_value: float = 1e-6) -> GradientTransformation:
+  """Rescale updates according to the Adam algorithm.
+
+  References:
+    [Zaheer et al, 2018](https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization)
```
Comment (mtthss) [inline optax/_src/transform.py:418]: Please user shorter link: [Zaheer et al, 2018](https://bit.ly/3nayVcT)

File: optax/__init__.py [hunk 4]
```diff
@@ -134,6 +134,7 @@
     "scale_by_schedule",
     "scale_by_stddev",
     "scale_by_trust_ratio",
+    "scale_by_yogi",
```
Comment (mtthss) [inline optax/__init__.py:137]: Also add actual python import higher up in the file

File: optax/__init__.py [hunk 5]
```diff
@@ -147,6 +148,7 @@
     "trace",
     "TraceState",
     "Updates",
+    "yogi",
```
Comment (mtthss) [inline optax/__init__.py:151]: Also add actual python import higher up in the file

File: optax/_src/alias.py [hunk 6]
```diff
@@ -109,3 +109,14 @@ def sgd(learning_rate: float,
       transform.trace(decay=momentum, nesterov=nesterov),
       transform.scale(-learning_rate),
   )
+
+
```
Comment (mtthss) [inline optax/_src/alias.py:113]: Use only 2 blank lines

File: optax/_src/alias.py [hunk 7]
```diff
@@ -109,3 +109,14 @@ def sgd(learning_rate: float,
...
+def yogi(learning_rate: float,
+              b1: float = 0.9,
+              b2: float = 0.999,
+              eps: float = 1e-3) -> GradientTransformation:
+  return combine.chain(
+      transform.scale_by_yogi(b1=b1, b2=b2, eps=eps),
+      transform.scale(-learning_rate),
+  )
```
Comment (mtthss) [inline optax/_src/alias.py:122]: leave 1 blank line at the end of file

**Comment by mtthss** [comment]
Left a few minor style comments, but code looks great!

**Comment by mtthss** [comment]
And apologies for the slow pace of review for your two last PRs.

**Comment by joaogui1** [comment]
I believe it's done!

File: optax/_src/transform.py [hunk 8]
```diff
@@ -408,6 +408,49 @@ def update_fn(updates, state, params=None):
...
+  References:
+    [Zaheer et al, 2018](https://papers.nips.cc/paper/2018/hash/90365351ccc7437a1309dc64e4db32a3-Abstract.html)
+
+  Args:
+    b1: decay rate for the exponentially weighted average of grads.
+    b2: decay rate for the exponentially weighted average of variance of grads.
+    eps: term added to the denominator to improve numerical stability.
+    eps_root: term added to the denominator inside the square-root to improve
+      numerical stability when backpropagating gradients through the rescaling.
+    initial_accumulator_value: The starting value for accumulators. Only positive values are allowed.
```
Comment (mtthss) [inline optax/_src/transform.py:426]: Please break line after `for accumulators.`

File: optax/_src/transform.py [hunk 9]
```diff
@@ -408,6 +408,49 @@ def update_fn(updates, state, params=None):
...
 
 
+def scale_by_yogi(
+    b1: float = 0.9, b2: float = 0.999,
+    eps: float = 1e-3, eps_root: float = 0.0, 
+    initial_accumulator_value: float = 1e-6) -> GradientTransformation:
+  """Rescale updates according to the Adam algorithm.
+
+  References:
+    [Zaheer et al, 2018](https://papers.nips.cc/paper/2018/hash/90365351ccc7437a1309dc64e4db32a3-Abstract.html)
```
Comment (mtthss) [inline optax/_src/transform.py:418]: Please add `#pylint:disable=line-too-long` after the link (on the same line)

**Comment by mtthss** [comment]
Apologies, last tiny changes and we are good to go!

**Comment by joaogui1** [comment]
No problem, my mistakes.

**Review by mtthss — APPROVED** [review]

File: optax/_src/alias.py [hunk 10]
```diff
@@ -109,3 +109,13 @@ def sgd(learning_rate: float,
       transform.trace(decay=momentum, nesterov=nesterov),
       transform.scale(-learning_rate),
   )
+
+
+def yogi(learning_rate: float,
+              b1: float = 0.9,
+              b2: float = 0.999,
+              eps: float = 1e-3) -> GradientTransformation:
```
Comment (mtthss) [inline optax/_src/alias.py:117]: There is still a formatting issue actually,
b1, b2, eps should be indented in line with learning_rate

**Comment by joaogui1** [comment]
We need a pylint config

**Review by mtthss — APPROVED** [review]

**Comment by mtthss** [comment]
PR is now submitted! Thanks for the contribution! :)

> We need a pylint config
Good suggestion, I will be adding one next week

**Pull request merged**